[Forge] update version to 1.0.7

9prady9 opened the pull request:
**Describe the pull request**
Update forge port to newer version

- What does your PR fix?
Port upgrade

- Which triplets are supported/not supported? Have you updated the CI baseline?
NA

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
- To the best of my knowledge.

I tried to use the following based on documentation from https://vcpkg.readthedocs.io/en/latest/README/ but some cmake functions with `vcpk_cmake` are being reported as not found on master branch. This is an unrelated problem though.

JonLiu1993 commented on ports/forge/vcpkg.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "name": "forge",
+  "version-string": "1.0.6",
```
@9prady9 ,When the version uses the A dot-separated sequence of numbers table,  we should use "version" instead of "version-string", the detailed document can refer to [here](https://github.com/microsoft/vcpkg/blob/master/docs/users/versioning.md#version)

9prady9 replied:
@JonLiu1993 Any idea why I am getting this after I changed `"version": 1.0.6`
```
MINGW64 /e/gitroot/vcpkg (forge_version_update)
$ ./vcpkg.exe x-add-version forge --overwrite-version
Error: Couldn't load port `forge`.
pradeep@cocanada MINGW64 /e/gitroot/vcpkg (forge_version_update)
$ ./vcpkg.exe  search forge
Error: while loading E:\gitroot\vcpkg\ports\forge\vcpkg.json:
E:\gitroot\vcpkg\ports\forge\vcpkg.json:3:17: error: Unexpected character; expected comma or close brace
   on expression:   "version": 1.0.6,
                                  ^

Error: failed to load port from E:\gitroot\vcpkg\ports\forge
Note: Updating vcpkg by rerunning bootstrap-vcpkg may resolve this failure.

MINGW64 /e/gitroot/vcpkg (forge_version_update)
$ ./vcpkg.exe x-add-version forge --overwrite-version
Error: Couldn't load port `forge`.
```

What is the overwrite-version command supposed to do ?

JonLiu1993 replied:
It seems that it is not allowed to be changed to "version", I will make changes on my local and reproduce this problem

9prady9 replied:
@JonLiu1993 My mistake, I expected the version to work without quotes. Once I added quotes around semver version, it worked fine.

Running the `./vcpkg x-add-version forge --overwrite-version` modified some additional files. Should those changes be a separate commit or should I include them along with my changed in a single commit ?

JonLiu1993 replied:
@9prady9 ,It should be submitted separately. After you modify your content, you make the first commit, and then you run the _./vcpkg x-add-version forge --overwrite-version_ command, then submit again, and finally unified git push finished

JonLiu1993 commented:
@9pradyd,After you have modified everything, you should run the ./vcpkg x-add-version forge --overwrite-version command and then submit it again

9prady9 commented:
@JonLiu1993 Took care of the feedback, please have a look again.

9prady9 commented:
@JonLiu1993 I am curious as to why `port-version` field is removed by `vcpkg format-manifest` command ?

9prady9 commented:
@JonLiu1993 

Found the problem that is happening on Azure
```
CMake Error at D:/downloads/tools/cmake-3.19.2-windows/cmake-3.19.2-win32-x86/share/cmake-3.19/Modules/ExternalProject.cmake:2542 (message):
  error: could not find git for clone of fg_glad-populate
Call Stack (most recent call first):
  D:/downloads/tools/cmake-3.19.2-windows/cmake-3.19.2-win32-x86/share/cmake-3.19/Modules/ExternalProject.cmake:3430 (_ep_add_download_command)
  CMakeLists.txt:13 (ExternalProject_Add)


-- Configuring incomplete, errors occurred!
See also "D:/buildtrees/forge/x64-windows-rel/extern/fg_glad-subbuild/CMakeFiles/CMakeOutput.log".
```
Forge 1.0.6 uses CMake's `FetchContent` style workflow to fetch couple of header only dependencies. Hence, it is looking for `git` program, is it forbidden to use FetchContent in vcpkg ports ? what is the suggested work around for this ?

9prady9 commented:
Can you please let me know how to circumvent the `git` not found problem that I shared in my earlier comment. Thank you.

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,20 +5,17 @@ endif()
     HEAD_REF master
-	PATCHES fix-static_build.patch
 )
 
```
```suggestion

vcpkg_find_acquire_program(GIT)

```

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,20 +5,17 @@ endif()
     SOURCE_PATH ${SOURCE_PATH}
-    PREFER_NINJA
     OPTIONS
         -DFG_BUILD_DOCS=OFF
```
```suggestion
        -DGIT_EXECUTABLE=${GIT}
        -DFG_BUILD_DOCS=OFF
```

9prady9 replied:
@JackBoosY Thanks for showing me how. I was under the impression tools like git that are often used by cmake are available by default.

JackBoosY commented on versions/f-/forge.json:
```diff
@@ -1,5 +1,10 @@
 {
   "versions": [
+    {
+      "git-tree": "d754c3fa314c2ddc3fe674fee7c11d611e5169df",
```
```suggestion
      "git-tree": "dcffe21bd11f2c1f35e85ddc9b67255f59be8e90",
```

9prady9 commented:
I looked into osx log files but there are only following files which doesn't give any clues to the cause of failure.

- VCPKGHeadersDatabase.txt
- VCPKGDatabase.txt

JackBoosY commented:
@9prady9  We'll fix this issue tomorrow.

jacobkahn commented:
@JonLiu1993 @JackBoosY — gentle bump on merging this assuming the above issue is fixed?

strega-nil commented on ports/forge/portfile.cmake:
```diff
@@ -5,20 +5,22 @@ endif()
 )
 
+vcpkg_find_acquire_program(GIT)
+get_filename_component(GIT_EXE_PATH ${GIT} DIRECTORY)
+vcpkg_add_to_path(${GIT_EXE_PATH})
```
Is there any need for these two lines given line 20?

9prady9 replied:
In fact it is the other way around, line 20 is not needed anymore. I didn't expect the need of this either but cmake's fetchcontent(internally ExternalProject_Add) command doesn't relay the variables like `GIT_EXECUTABLE` automatically to the sub-project. 

This problem will be there for all projects that use FetchContent from CMake.

9prady9 commented:
> @JonLiu1993 @JackBoosY — gentle bump on merging this assuming the above issue is fixed?

Actually, I have improved cmake support in the port. If that is alright, I would like to put this aside for about a week or so to give me time to push another release for forge with better static library build experience and another code bug.

9prady9 commented:
@JackBoosY @JonLiu1993 I have updated forge port to latest release version. I don't think failure in CI job is related to forge based on the logs. Any idea what is the cause of the failure ?

JonLiu1993 commented:
@9prady9 ,Depends on #17528

JackBoosY commented:
Waiting for #17528 merge.

9prady9 commented:
> Waiting for #17528 merge.

Ah, okay. Thank you for the quick response :)

9prady9 commented:
A quick question on how vcpkg handles the following line

https://github.com/microsoft/vcpkg/blob/ec9579b7219fe5c79bd005ae8541460c5bb43233/ports/forge/portfile.cmake#L26

Despite the line having `share/Forge/cmake` the word forge is changing to lower case when the cmake config files are installed via vcpkg. What am I missing here ?

The problem with this name difference is that config file is forge is ending being not found.

Any hints on how to resolve this are appreciated. Thank you. Kindly let me know before merging this PR in. I want to push this fix also here itself which should wrap up my changes.

9prady9 commented:
> A quick question on how vcpkg handles the following line
> 
> https://github.com/microsoft/vcpkg/blob/ec9579b7219fe5c79bd005ae8541460c5bb43233/ports/forge/portfile.cmake#L26
> 
> Despite the line having `share/Forge/cmake` the word forge is changing to lower case when the cmake config files are installed via vcpkg. What am I missing here ?
> 
> The problem with this name difference is that config file is forge is ending being not found.
> 
> Any hints on how to resolve this are appreciated. Thank you. Kindly let me know before merging this PR in. I want to push this fix also here itself which should wrap up my changes.

nvm, I found the docs for fixup function. Please ignore earlier question.

9prady9 commented:
I do have a question about new commands though `vcpkg_cmake_config_fixup` why do they often cause unknown command error ? Should I enable anything specific to use newer functions ?

9prady9 commented:
> > A quick question on how vcpkg handles the following line
> > https://github.com/microsoft/vcpkg/blob/ec9579b7219fe5c79bd005ae8541460c5bb43233/ports/forge/portfile.cmake#L26
> > 
> > Despite the line having `share/Forge/cmake` the word forge is changing to lower case when the cmake config files are installed via vcpkg. What am I missing here ?
> > The problem with this name difference is that config file is forge is ending being not found.
> > Any hints on how to resolve this are appreciated. Thank you. Kindly let me know before merging this PR in. I want to push this fix also here itself which should wrap up my changes.
> 
> nvm, I found the docs for fixup function. Please ignore earlier question.

I  am seeing some odd behavior with `vcpkg_fixup_cmake_targets` cmake command.

From vcpkg repository, it works as expected - I get a `share/Forge/cmake` for my installation cmake files. However, when I use the same local repo's `vcpkg.cmake` file in another repository which has a `vcpkg.json` manifest file. I still get the old `share/forge/cmake` path. I have removed the cached files under `~/.cache` and tried also. Can you please help me figure out what is going on here.

9prady9 commented:
> > > A quick question on how vcpkg handles the following line
> > > https://github.com/microsoft/vcpkg/blob/ec9579b7219fe5c79bd005ae8541460c5bb43233/ports/forge/portfile.cmake#L26
> > > 
> > > Despite the line having `share/Forge/cmake` the word forge is changing to lower case when the cmake config files are installed via vcpkg. What am I missing here ?
> > > The problem with this name difference is that config file is forge is ending being not found.
> > > Any hints on how to resolve this are appreciated. Thank you. Kindly let me know before merging this PR in. I want to push this fix also here itself which should wrap up my changes.
> > 
> > 
> > nvm, I found the docs for fixup function. Please ignore earlier question.
> 
> @JackBoosY @JonLiu1993 I am seeing some odd behavior with `vcpkg_fixup_cmake_targets` cmake command.
> 
> From vcpkg repository, it works as expected - I get a `share/Forge/cmake` for my installation cmake files. However, when I use the same local repo's `vcpkg.cmake` file in another repository which has a `vcpkg.json` manifest file. I still get the old `share/forge/cmake` path. I have removed the cached files under `~/.cache` and tried also. Can you please help me figure out what is going on here.

I finally figured it out - without doing `vcpkg x-add-version forge ...` after there are any changes to either the manifest or portfile, changes are not reflected properly. Earlier, I was under the impression that this is true for changes in manifest file only.

In any case, it is kind of still confusing how installation in vcpkg works but not if it is done inside project build folder using manifest file.

9prady9 commented:
Phew, finally I think I am done with the change. As it turned out `vcpkg_fixup_cmake_targets` was not needed at all. I wonder why it was originally added.

@JonLiu1993 @JackBoosY I think this is ready for final review/merge. Thank you for working with me on getting this updated.

9prady9 commented:
@JackBoosY I see that you separately converted forge to manifest format. The usage patch is not needed if the fixup call is removed. And that call is not needed either. Are you okay if I remove it ?

JackBoosY commented:
Can you please merge to master?
And, `vcpkg_fixup_cmake_targets` is required because it fixes some absoultly path in cmake configure files.

9prady9 commented:
> Can you please merge to master?
> And, `vcpkg_fixup_cmake_targets` is required because it fixes some absoultly path in cmake configure files.

I have tested this without `vcpkg_fixup_cmake_targets` and seems to be working fine. Have you tried without it ? Having that call is causing path issues that I am unable to fix by any argument to that function.

I can push the changes I have and you can perhaps have a look at those once.

JackBoosY commented:
@9prady9 In the cmake configuration file, you can find some absolute paths. If the user exports this port, it will not be available.

9prady9 commented:
> 
> 
> @9prady9 In the cmake configuration file, you can find some absolute paths. If the user exports this port, it will not be available.

@JackBoosY Here is the problem I am facing. With the following 
```cmake
    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake)
```
the TARGET_PATH defaults to `share/${PORT}` and since port name is different from package name. This causes issue when this packages `ForgeConfig.cmake` is used by any user of this port. This obviously can't be the solution until a point where forge project's package name and port name match.

Now, if I change it to the following in order to address the port name and package name difference.
```cmake
vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake TARGET_PATH share/Forge/cmake)
```
now, the `PACKAGE_PREFIX_DIR` that is computed by cmake inside the `ForgeConfig.cmake` file ends up pointing to `build/vcpkg_installed/x64-linux/share` instead of `build/vcpkg_installed/x64-linux`. Note that this happens when I use this PR's branch of vcpkg in my local arrayfire project to use vcpkg's locally updated port. (When I install `forge` inside the vcpkg repository folder, folder heirarchy looks fine.) 

Now, this is when I referred to existing [arrayfire port](https://github.com/microsoft/vcpkg/blob/master/ports/arrayfire/portfile.cmake) to see what it is doing. Removing this `vcpkg_fixup_cmake_targets` actually installed the `forge` port fine both inside the vcpkg repository and in a project's build directory. I was also able to build arrayfire using the installed vcpkg forge. I am able to use this port in another project but you claim otherwise. Therefore, I am confused.

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,30 +5,21 @@ endif()
 )
 
 vcpkg_install_cmake()
 
```
```suggestion
if (VCPKG_TARGET_IS_WINDOWS)
    vcpkg_fixup_cmake_targets(CONFIG_PATH cmake TARGET_PATH share/Forge)
else()
    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake TARGET_PATH share/Forge)
endif()
```

JackBoosY replied:
In the official [documentation](https://cmake.org/cmake/help/latest/command/find_package.html), they said:
Config mode search attempts to locate a configuration file provided by the package to be found. A cache entry called  ${PackageName}_DIR is created to hold the directory containing the file. By default the command searches for a package with the name PackageName. If the NAMES option is given the names following it are used instead of PackageName 
**The command searches for a file called ${PackageNameConfig}.cmake or  ${lower-case-package-name}-config.cmake for each name specified**.

JackBoosY commented:
@9prady9 Please see:
1. https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure
2. https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L482-L508

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,30 +5,21 @@ endif()
 
 vcpkg_install_cmake()
 
-if (VCPKG_TARGET_IS_WINDOWS)
-    vcpkg_fixup_cmake_targets(CONFIG_PATH cmake)
-else()
-    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake)
-endif()
```
Please don't remove them.

9prady9 commented:
> "Please don't remove them"

@JackBoosY I am not trying to remove anything that is needed. I have clearly explained [here](https://github.com/microsoft/vcpkg/pull/17099#issuecomment-830169641) the issues that I am facing while using the following.
```cmake
if (VCPKG_TARGET_IS_WINDOWS)
    vcpkg_fixup_cmake_targets(CONFIG_PATH cmake)
else()
    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake)
endif()
```
I considered removing it as fix because there were portfiles in other ports(like arrayfire) that don't call this `vcpkg_fixup_cmake_targets` which made me think it isn't mandatory for every cmake based project. Kindly correct me if that is wrong. Can you please clarify why the above call is required in some ports and not others.

>    https://cmake.org/cmake/help/latest/command/find_package.html#search-procedure
>   https://github.com/microsoft/vcpkg/blob/master/scripts/buildsystems/vcpkg.cmake#L482-L508
> In all cases the <name> is treated as case-insensitive and corresponds to any of the names specified (<PackageName> or names given by NAMES).

I don't think problem is CMake or an incorrect name Forge's CMake. Please go through the comment https://github.com/microsoft/vcpkg/pull/17099#issuecomment-830169641 . I can share a sample project with forge added as dependency in it's manifest if you think that can help you reproduce the issue on your end.

>    You should use vcpkg toolchain.
>    If not, you shoud set Forge_DIR to EXPORT_PATH/share/forge.

Yes, I am aware of both these procedures. I am already using toolchain approach along with a vcpkg manifest in project root.
```cmake
cmake .. -DCMAKE_TOOLCHAIN_FILE:FILEPATH=<vcpkg.cmake file path>
```
That is only thing that needs to be done for automatically installation of dependencies and later to be found by `find_package` calls. I am already using this.

9prady9 reviewed (commented):
@JackBoosY please check my response here https://github.com/microsoft/vcpkg/pull/17099#issuecomment-830241185

9prady9 commented:
Finally figured out a work around for the earlier problem https://github.com/microsoft/vcpkg/pull/17099#issuecomment-830241185 and also added a patch for forge's cmake config when multiple build configurations are involved.

@JackBoosY Setting the `FG_INSTALL_CMAKE_DIR` helped work around the path inconsistencies that were happening.

JackBoosY commented:
@9prady9  Let me explain more clearly:
1. In vcpkg, all cmake configuration files(`*Config.cmake` / `*-config.cmake` / `*Targets.cmake` / `*Targets-debug.cmake` / `*Targets-release.cmake`) should be installed in _share/${PORT}_.
2. `vcpkg_fixup_cmake_targets` not only fixes the absolute path, but also moves the cmake debug configuration file to _share/${PORT}_.
3. If the generated file name is different from the port name `${PORT}`, we need to add the parameter `TARGET_PATH` to `vcpkg_fixup_cmake_targets` to correct its folder name(`${FOLDERNAME-config.cmake}`/ `${FOLDERNAME}Config.cmake`).

JackBoosY commented:
And for the issue you met, I've already fixed that in PR https://github.com/microsoft/vcpkg/tree/master/ports/forge.
Please merge to master and apply all patches.

Thanks.

9prady9 commented:
> And for the issue you met, I've already fixed that in PR #17404.
> Please merge to master and apply all patches.
> 
> Thanks.
> 
> Edit: I noticed you removed my patch, please revert it.

@JackBoosY Your patch is not removed totally. I added a single line that addresses on both Windows and non-Windows.

I thought I already rebased it to the latest when I commented here.

JackBoosY commented on ports/forge/vcpkg.json:
```diff
@@ -1,8 +1,8 @@
-  "port-version": 1,
-  "description": "Helps with high performance visualizations involving OpenGL-CUDA/OpenCL interop.",
+  "version-semver": "1.0.7",
+  "port-version": 0,
```
```suggestion
```
Yes, please remove this line.

9prady9 replied:
Okay. yes, this seems to be done in past couple of days.

9prady9 commented:
@JackBoosY Finally, green, pls have a look.

JackBoosY commented on ports/forge/forge_cmake_multiconfig.patch:
```diff
@@ -0,0 +1,36 @@
+diff --git a/CMakeModules/ForgeConfig.cmake.in b/CMakeModules/ForgeConfig.cmake.in
+index e4c7737..6fb18ac 100644
+--- a/CMakeModules/ForgeConfig.cmake.in
++++ b/CMakeModules/ForgeConfig.cmake.in
```
Forge link to `OpenGL::GL` but missing to find the depencency `glad`, please add the following codes to this file:
```
include(CMakeFindDependencyMacro)
find_depencency(glad CONFIG)
```
when include the targets files.

9prady9 replied:
Yes, `find_dependency` checks are overdue. But I think such checks to be done for all upstream dependencies, just `glad` won't be sufficient. I will update the patch accordingly.

9prady9 replied:
@JackBoosY I have just pushed the changes, please have a look.

JackBoosY commented on ports/forge/cmake_config.patch:
```diff
@@ -0,0 +1,58 @@
++    # Check for dependencies when Forge binary and cmake files are found
++    include(CMakeFindDependencyMacro)
++
++    find_dependency(glad CONFIG REQUIRED)
++    find_dependency(glfw3 CONFIG REQUIRED)
++    find_dependency(freetype CONFIG REQUIRED)
++    find_dependency(freeimage CONFIG REQUIRED)
```
When we use `find_dependency`, we don't need to add keyword `REQUIRED`.

JackBoosY replied:
The target `OpenGL::GL` is added in _CMakeModules\FindOpenGL.cmake_, I think you should copy it to _share/Forge_, and add `find_dependency(OpenGL)` here.
I checked, it works.

9prady9 replied:
> When we use `find_dependency`, we don't need to add keyword `REQUIRED`.

@JackBoosY What is the default behavior if keyword is not specified ? It is not incorrect to specify `REQUIRED`, what is your concern here may I know ?

9prady9 replied:
> The target `OpenGL::GL` is added in _CMakeModules\FindOpenGL.cmake_, I think you should copy it to _share/Forge_ (write it to `portfile.cmake`), and add `find_dependency(OpenGL)` here.
> I checked, it works.

Just `find_dependency(OpenGL)` is sufficient I think. Why do you need a copy of that Find Module ? FindOpenGL is already provided by cmake modules.

JackBoosY replied:
@9prady9  The default is `REQUIRED`, so we don't need add it anymore.
Obviously, forge uses the internal `FindOpenGL.cmake`, if it can be successfully built by deleting it before build, we don’t need to install it in share folder anymore.

9prady9 replied:
So, removing `REQUIRED` from all `find_dependency` calls is sufficient, correct ?

JackBoosY replied:
Yes.

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,31 +5,23 @@ endif()
-        fix-usage.patch
+    PATCHES cmake_config.patch
 )
 
```
```suggestion

file(REMOVE ${SOURCE_PATH}/CMakeModules/FindOpenGL.cmake)

```

JackBoosY replied:
Please also remove this internal cmake file.

9prady9 replied:
Is this so that, cmake's own module is found by `find_package(OpenGL)` call ?

JackBoosY replied:
@9prady9 cmake will give priority to the search components in the source directory, rather than cmake's own.

9prady9 replied:
I don't think [CMAKE_MODULE_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_MODULE_PATH.html) works like that. It is empty by default even in any CMake project. It has to be set explicitly after project call. For given set of installed cmake files which doesn't add any special path to it, CMake will only search it's own default set of Modules. What am I missing from the documentation ? I haven't seen this behavior before, hence my question.

9prady9 replied:
@JackBoosY Do you agree with my assessment ? Or is there something that cmake/vcpkg is doing behind the scenes which I am not aware of ?

JackBoosY replied:
Sadly, Iit first searches forge's own cmake module, I checked.

9prady9 replied:
Hm, must be a bug in cmake perhaps

Unfortunate that such source modifications are necessary.  Pushed the updates. please check now

JackBoosY commented on ports/forge/cmake_config.patch:
```diff
@@ -0,0 +1,59 @@
++            message(FATAL_ERROR "Please install Fontconfig dependency")
++        endif()
++    endif()
++endif()
```
```
CMake Error in F:/vcpkg/installed/x64-windows/share/Forge/ForgeConfig.cmake:
  A logical block opening on the line
    F:/vcpkg/installed/x64-windows/share/Forge/ForgeConfig.cmake:111 (if)
  is not closed.
Call Stack (most recent call first):
  F:/vcpkg/scripts/buildsystems/vcpkg.cmake:862 (_find_package)
  CMakeProject2/CMakeLists.txt:141 (find_package)
```
I checked the config file from ln 111 to the end:
```
if(Forge_FOUND)
    # Check for dependencies when Forge binary and cmake files are found
    include(CMakeFindDependencyMacro)

    find_dependency(glad CONFIG)
    find_dependency(glfw3 CONFIG)
    find_dependency(freetype CONFIG)
    find_dependency(freeimage CONFIG)
    find_dependency(OpenGL)

    if(NOT WIN32)
        find_package(PkgConfig REQUIRED)
        pkg_check_modules(fontconfig REQUIRED)
        if(NOT fontconfig_FOUND)
            message(FATAL_ERROR "Please install Fontconfig dependency")
        endif()
    endif()

```
Maybe your patch has some problems, please re-make it.

9prady9 replied:
I know what is happening, bah. that is result of manual patch edit that I did. Let me fix that quickly.

9prady9 replied:
Should be fixed now, sorry about that.

JackBoosY commented on ports/forge/portfile.cmake:
```diff
@@ -5,32 +5,25 @@ endif()
-    vcpkg_fixup_cmake_targets(CONFIG_PATH share/Forge/cmake)
-endif()
-
+file(REMOVE ${SOURCE_PATH}/CMakeModules/FindOpenGL.cmake)
```
Please move this line to line 13.

9prady9 replied:
what is the point of it ? this is source path right ?

JackBoosY replied:
@9prady9 We should delete it before configuring cmake, otherwise it will be used.

9prady9 replied:
Wait, I think I misunderstood your earlier explanation. I was under the impression that you are saying somehow any cmake project that uses vcpkg's forge port will use this FindOpenGL instead of default system find modules. That is why it didn't make sense to me earlier when you said cmake is using this.

Are you talking about forge's build processing using this custom `FindOpenGL` module ?

JackBoosY replied:
This makes sense, because the FindOpenGL is provided by the upstream (forge), and there is no guarantee that it is consistent with the one provided by cmake.
Since it was NOT deleted before configuration, the configuration process will use it and add the results (marcos of libraries and header files) to the cmake configuration file of forge(Forge-targets.cmake).
Once cmake modifies `FindOpenGL.cmake`, when the user uses `ForgeConfig.cmake`, `find_dependency(OpenGL)` in it uses the updated `FindOpenGL.cmake` in cmake, which results in mismatch of macros or other things, which may cause usage problems.
This problem can be avoided by deleting it before configuration to use the same FindOpenGL during configuration and use.

JackBoosY replied:
See step:
Assuming that the two `FindOpenGL.cmake` (one for forge and the other for cmake) do not match.
Build: Configure -> `FindOpenGL.cmake` in forge -> build ->export the cmake files -> install
Use: Call `find_package(Forge CONFIG REQUIRED)` -> `find_dependency(OpenGL)` -> `FindOpenGL.cmake` in cmake -> may caused error.

9prady9 replied:
Okay, I understand the scenario you explained earlier clearly now.

Here's what is going on with Forge, our minimum required cmake version is 3.5 and FindOpenGL has changed quite drastically since that version and 3.20, current release. I picked a version of this module from particular cmake version to be consistent in our build process across platforms. Removing this file will ensure modules from vcpkg download cmake are picked up. Fine by me as long as it has required OpenGL cmake targets.

As vcpkg doesn't use older cmake versions, I guess there is no issue on that front. I was trying to back port any fixes I am adding in this patch in the upstream project. So, I was trying to understand clearly what issues you were experiencing/expecting, so that I don't end up breaking usage for old cmake versions. Thanks for the detailed explanation.

Updated the PR.

JackBoosY commented on ports/forge/cmake_config.patch:
```diff
@@ -0,0 +1,58 @@
++    find_dependency(glfw3 CONFIG REQUIRED)
++    find_dependency(freetype CONFIG REQUIRED)
++    find_dependency(freeimage CONFIG REQUIRED)
++
```
Why you removed `find_dependency(OpenGL)` here?

9prady9 replied:
ahhh, I didn't intentionally. stashed diff might not have had it. Give me a minute.

9prady9 replied:
@JackBoosY please check now

9prady9 committed:
[forge] Update forge to v1.0.7 tag

- Use explicit cmake files install path
- Use REF option in vcpkg_from_github command to fetch tag

9prady9 committed:
[forge] Update version in baseline

9prady9 commented:
Can you please let me know the ETA for this change to be available in master

JackBoosY commented:
@9prady9  Soon.
Ping @dan-shaw for merge this PR.
